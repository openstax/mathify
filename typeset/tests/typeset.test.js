const path = require('path')
require('dotenv').config()
const fs = require('fs')
const fileExists = require('file-exists')
const bunyan = require('bunyan')
const BunyanFormat = require('bunyan-format')
const converter = require('./../converter')
const { createHash } = require('crypto')

const log = bunyan.createLogger({
  name: 'node-typeset',
  level: process.env.LOG_LEVEL || 'info',
  stream: new BunyanFormat({ outputMode: process.env.LOG_FORMAT || 'short' })
})

const pathToInput = path.resolve('./typeset/tests/seed/test.baked.xhtml')
const pathToCss = path.resolve('./typeset/tests/seed/test.css')
const pathToOutput = path.resolve('./typeset/tests/test-output.xhtml')
const pathToOutputSVG = path.resolve('./typeset/tests/test-output-svg.xhtml')
const pathToInputLatex = path.resolve('./typeset/tests/seed/test-latex.xhtml')
const pathToInputLatexErrors = path.resolve('./typeset/tests/seed/test-latex-errors.xhtml')
const pathToOutputLatex = path.resolve('./typeset/tests/test-output-latex.xhtml')
const pathToOutputMML = path.resolve('./typeset/tests/test-output-mathml.xhtml')

const pathToCodeInput = path.resolve('./typeset/tests/seed/test-code.xhtml')
const pathToCodeOutput = path.resolve('./typeset/tests/test-code.output.xhtml')

beforeAll(() => {
  if (fileExists.sync(pathToOutput)) {
    fs.unlink(pathToOutput, (err) => {
      if (err) throw err
    })
  }

  if (fileExists.sync(pathToOutputSVG)) {
    fs.unlink(pathToOutputSVG, (err) => {
      if (err) throw err
    })
  }

  if (fileExists.sync(pathToOutputLatex)) {
    fs.unlink(pathToOutputLatex, (err) => {
      if (err) throw err
    })
  }

  if (fileExists.sync(pathToOutputMML)) {
    fs.unlink(pathToOutputMML, (err) => {
      if (err) throw err
    })
  }

  if (fileExists.sync(pathToCodeOutput)) {
    fs.unlink(pathToCodeOutput, (err) => {
      if (err) throw err
    })
  }
})

afterAll(() => {
  if (fileExists.sync(pathToOutput)) {
    fs.unlink(pathToOutput, (err) => {
      if (err) throw err
    })
  }

  if (fileExists.sync(pathToOutputSVG)) {
    fs.unlink(pathToOutputSVG, (err) => {
      if (err) throw err
    })
  }

  if (fileExists.sync(pathToOutputLatex)) {
    fs.unlink(pathToOutputLatex, (err) => {
      if (err) throw err
    })
  }

  if (fileExists.sync(pathToOutputMML)) {
    fs.unlink(pathToOutputMML, (err) => {
      if (err) throw err
    })
  }

  if (fileExists.sync(pathToCodeOutput)) {
    fs.unlink(pathToCodeOutput, (err) => {
      if (err) throw err
    })
  }
})

function getHashFile (fpath) {
  return new Promise((resolve, reject) => {
    const hash = createHash('sha256')
    const reader = fs.createReadStream(fpath).setEncoding('utf8')
    reader.on('data', chunk => hash.update(chunk))
    reader.on('error', err => reject(err))
    reader.on('end', () => {
      resolve(hash.digest('hex'))
    })
  })
}

test('Fail if user provide wrong path for input file (Math).', async (done) => {
  const res = await converter.createMapOfMathMLElements(log, './wrong/path.xhtml', pathToCss, pathToOutput, 'html', 3000, true)
  expect(res).toBe(converter.STATUS_CODE.ERROR)
  done()
})

test('Fail if user provide wrong path for css file.', async (done) => {
  const res = await converter.createMapOfMathMLElements(log, pathToInput, './wrong/path.xhtml', pathToOutput, 'html', 3000, true)
  expect(res).toBe(converter.STATUS_CODE.ERROR)
  done()
})

test('Success if converter finished without errors FORMAT HTML.', async (done) => {
  const res = await converter.createMapOfMathMLElements(log, pathToInput, pathToCss, pathToOutput, 'html', 3000, true)
  let isOutputFile = false
  if (fileExists.sync(pathToOutput)) {
    isOutputFile = true
  }
  expect(res).toBe(converter.STATUS_CODE.OK)
  expect(isOutputFile).toBeTruthy()
  expect(await getHashFile(pathToOutput)).toMatchSnapshot()
  done()
}, 30000)

test('Success if converter finished without errors FORMAT SVG.', async (done) => {
  const res = await converter.createMapOfMathMLElements(log, pathToInput, pathToCss, pathToOutputSVG, 'svg', 3000, true)
  let isOutputFile = false
  if (fileExists.sync(pathToOutputSVG)) {
    isOutputFile = true
  }
  expect(res).toBe(converter.STATUS_CODE.OK)
  expect(isOutputFile).toBeTruthy()
  expect(await getHashFile(pathToOutputSVG)).toMatchSnapshot()
  done()
}, 30000)

test('Success if convertered LaTeX functions with success.', async (done) => {
  const res = await converter.createMapOfMathMLElements(log, pathToInputLatex, pathToCss, pathToOutputLatex, 'html', 3000, true)
  let isOutputFile = false
  if (fileExists.sync(pathToOutputLatex)) {
    isOutputFile = true
  }
  expect(res).toBe(converter.STATUS_CODE.OK)
  expect(isOutputFile).toBeTruthy()
  expect(await getHashFile(pathToOutputLatex)).toMatchSnapshot()

  done()
}, 30000)

test('Success if convertered LaTeX to mathml with success.', async (done) => {
  const res = await converter.createMapOfMathMLElements(log, pathToInputLatex, pathToCss, pathToOutputMML, 'mathml', 3000, true)
  let isOutputFile = false
  if (fileExists.sync(pathToOutputMML)) {
    isOutputFile = true
  }
  expect(res).toBe(converter.STATUS_CODE.OK)
  expect(isOutputFile).toBeTruthy()
  expect(await getHashFile(pathToOutputMML)).toMatchSnapshot()

  done()
}, 30000)

test('Convert inline code tags and block pre tags', async (done) => {
  const res = await converter.createMapOfMathMLElements(log, pathToCodeInput, pathToCss, pathToCodeOutput, 'html', 3000, true)
  expect(fileExists.sync(pathToCodeOutput))
  expect(res).toBe(converter.STATUS_CODE.OK)
  expect(fs.readFileSync(pathToCodeOutput, 'utf-8')).toMatchSnapshot()
  done()
}, 3000)

test('Error logging', async (done) => {
  jest.spyOn(console, 'error').mockImplementation(() => {})
  const messages = []
  const logCapture = {
    info: jest.fn(),
    debug: jest.fn(),
    error: jest.fn().mockImplementation(messages.push.bind(messages))
  }
  await expect(
    converter.createMapOfMathMLElements(logCapture, pathToInputLatexErrors, pathToCss, pathToOutputMML, 'mathml', 3000, true)
  ).rejects.toThrow()

  expect(messages).toStrictEqual(
    [
`{
  \"errors\": [
    \"TeX parse error: Undefined control sequence \\\\3\"
  ],
  \"tagName\": \"span\",
  \"data-sm\": \"./some-document\",
  \"data-math\": \"\\\\left\\\\{\\\\begin{array}{l}3x+7y=15\\\\3x+7y=40\\\\end{array}\\\\right.\",
  \"data-injected-from-nickname\": \"broken-exercise\"
}`,
`{
  \"errors\": [
    \"TeX parse error: Missing close brace\"
  ],
  \"tagName\": \"span\",
  \"data-sm\": \"./some-document\",
  \"data-math\": \"\\\\text{H^+\",
  \"data-injected-from-nickname\": \"broken-exercise\"
}`
    ]
  )

  jest.restoreAllMocks()
  done()
}, 30000)
