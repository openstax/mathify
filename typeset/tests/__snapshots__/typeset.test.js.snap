// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Convert inline code tags and block pre tags 1`] = `
"<html xmlns=\\"http://www.w3.org/1999/xhtml\\" xmlns:m=\\"http://www.w3.org/1998/Math/MathML\\">
  <body itemscope=\\"itemscope\\" itemtype=\\"http://schema.org/Book\\">

<pre class=\\"line-numbering\\" data-lang=\\"python\\"><table class=\\"hljs-ln\\"><tr><td class=\\"hljs-ln-line hljs-ln-numbers\\" data-line-number=\\"1\\"><div class=\\"hljs-ln-n\\" data-line-number=\\"1\\">1</div></td><td class=\\"hljs-ln-line hljs-ln-code\\" data-line-number=\\"1\\">my_integer = <span class=\\"hljs-number\\">5</span></td></tr><tr><td class=\\"hljs-ln-line hljs-ln-numbers\\" data-line-number=\\"2\\"><div class=\\"hljs-ln-n\\" data-line-number=\\"2\\">2</div></td><td class=\\"hljs-ln-line hljs-ln-code\\" data-line-number=\\"2\\">  my_floating_point = <span class=\\"hljs-number\\">26.2</span></td></tr><tr><td class=\\"hljs-ln-line hljs-ln-numbers\\" data-line-number=\\"3\\"><div class=\\"hljs-ln-n\\" data-line-number=\\"3\\">3</div></td><td class=\\"hljs-ln-line hljs-ln-code\\" data-line-number=\\"3\\">  my_Boolean = <span class=\\"hljs-literal\\">True</span></td></tr><tr><td class=\\"hljs-ln-line hljs-ln-numbers\\" data-line-number=\\"4\\"><div class=\\"hljs-ln-n\\" data-line-number=\\"4\\">4</div></td><td class=\\"hljs-ln-line hljs-ln-code\\" data-line-number=\\"4\\">  my_string = <span class=\\"hljs-string\\">'characters'</span></td></tr></table></pre>

<pre lang=\\"python\\">
  print(<span class=\\"hljs-string\\">'Hi'</span>)
  a = b + c
</pre>

<p>
  <code class=\\"line-numbering\\" data-lang=\\"python\\"><table class=\\"hljs-ln\\"><tr><td class=\\"hljs-ln-line hljs-ln-numbers\\" data-line-number=\\"1\\"><div class=\\"hljs-ln-n\\" data-line-number=\\"1\\">1</div></td><td class=\\"hljs-ln-line hljs-ln-code\\" data-line-number=\\"1\\">hello = <span class=\\"hljs-number\\">5</span></td></tr></table></code>
</p>

<p>
  <code data-lang=\\"python\\">hello = <span class=\\"hljs-number\\">5</span></code>
</p>

<p>
  <code lang=\\"python\\">print(<span class=\\"hljs-number\\">5</span>)</code>
</p>

</body>
</html>
"
`;

exports[`Success if converter finished without errors FORMAT HTML. 1`] = `"4ec8f2f5d38cb84f0a1a90013657292f6a53ffcfd6e0c0ce07ed13faf3e4d289"`;

exports[`Success if converter finished without errors FORMAT SVG. 1`] = `"9102725791b223ab8e8945097fe8050603d964edb0ed1bf17868cae9e9a85d41"`;

exports[`Success if convertered LaTeX functions with success. 1`] = `"1597bf77d3971a31fcd50daef5654847946ac512a4b63b397fe25c4951bfba75"`;

exports[`Success if convertered LaTeX to mathml with success. 1`] = `"f5744bc949145782349ecd1e9a0ea1376b23ddf74c585b3592004854c2b87dea"`;
